examples/linear_algebra.dx

```diff
@@ -0,0 +1,196 @@
+  -- Computes lower, upper, and permuntation matrices from a square matrix,
+  -- such that apply_permutation permutation a == lower ** upper.
+  -- I.e. it uses partial pivoting.
+  (permutation, swapcount) = pivotize a
+  a = apply_permutation permutation a
+
+  init_lower = for i:n. for j':(..i).
+    select (i == (%inject j')) 1.0 0.0
+  init_upper = for i:n. for j'':(i..). 0.0
+
+  (lower, upper) = snd $ withState (init_lower, init_upper) \stateRef.
+    lRef = fstRef stateRef
+    uRef = sndRef stateRef
+
+  -- For reference, here's code to computed the LU decomposition
+  -- without dependent tables (i.e. with standard flat matrices):
+  --  for j:n.
+  --    for i:(..j).
+  --      i = %inject i
+  --      s = sum for k':(..i).
+  --        k = %inject k'
+  --        (get uRef!k!j) * (get lRef!i!k)
+  --      uRef!i!j := a.i.j - s
+
+  --    for i':(j<..).
+  --      i = %inject i'
+  --      s = sum for k':(..j).
+  --        k = %inject k'
+  --        (get uRef!k!j) * (get lRef!i!k)
+  --      lRef!i!j := (a.i.j - s) / (get uRef!j!j)
+  --    for i:n. ()
+
+    for j:n.
+      for i:(..j).
+        i' = %inject i
+        s = sum for k:(..i).
+          k' = (ordinal k)@n
+          ukj = get (upperTriIndex uRef k')!(((ordinal j) - (ordinal k))@_)
+          lik = get (lowerTriIndex lRef i')!((ordinal k)@_)
+          ukj * lik
+          
+        uijRef = (upperTriIndex uRef i')!(((ordinal j) - (ordinal i))@_)
+        uijRef := a.(%inject i).j - s
+      
+      for i:(j<..).
+        i' = %inject i
+        s = sum for k:(..j).
+          k' = %inject k
+          i'' = (((ordinal j) - (ordinal k))@_)
+          ukj = get (upperTriIndex uRef k')!i''
+          lik = get (lowerTriIndex lRef i')!((ordinal k)@_)
+          ukj * lik
+        
+        i'' = ((ordinal i) + (ordinal j) + 1)@_
+        ujj = get (upperTriIndex uRef j)!(0@_)
+        lijRef = (lowerTriIndex lRef i'')!((ordinal j)@_)
+        lijRef := (a.i'.j - s) / ujj
+  (lower, upper, permutation, swapcount)
+
+
+'### General linear algebra functions.
+
+def solve (_:Eq n) ?=> (_:VSpace v) ?=> (a:n=>n=>Float) (b:n=>v) : n=>v =
+  -- There's a small speedup possible by exploiting the fact
+  -- that l always has ones on the diagonal.  It would just require a
+  -- custom forward_substitute routine that doesn't divide
+  -- by the diagonal entries.
+  (l, u, perm, _) = lu a
+  b' = apply_permutation perm b
+  y = forward_substitute l b'
+  backward_substitute u y
+
+def invert (_:Eq n) ?=> (a:n=>n=>Float) : n=>n=>Float =
+  solve a identity_matrix
+
+def determinant (_:Eq n) ?=> (a:n=>n=>Float) : Float =
+  (l, u, p, swapcount) = lu a
+  -- formerly u.i.i * l.i.i
+  (prod for i. u.i.(0@_) * l.i.((ordinal i)@_)) * (permutation_sign swapcount)
+
+def sign_and_log_determinant (_:Eq n) ?=> (a:n=>n=>Float) : (Float & Float) =
+  (l, u, p, swapcount) = lu a
+  -- formerly u.i.i * l.i.i
+  diags = for i. u.i.(0@_) * l.i.((ordinal i)@_)
```

Comment:
It seems like a lot of casts are to access the first or last elements of each row. Maybe we should just have the (partial) functions `first : n=>a -> a` and `last : n=>a -> a`?